### PR TITLE
Increment lclVar refCounts in fgMorphBlockStmt.

### LINKED ISSUE #8258: RyuJIT/x86: CoreFx System.Dynamic.Runtime.Tests Assertion failed '!"We should never encounter a reference to a lclVar that has a zero refCnt."
Commits:
dbf8ce1173faf78b176771946b0bca9f9863bc9b
dotnet/corefx@0d623b5b5e110a336eb743b6d9066f336a5177b4

Repro:
coreclr *(assuming root=c:\git\coreclr)*
```
build x86 checked skiptests
```
corefx *(assuming root=c:\git\corefx)*
```
build -Release -SkipTests -- /p:BUILDTOOLS_OVERRIDE_RUNTIME=c:\git\coreclr\bin\Product\Windows_NT.x64.Checked
cd bin\tests\AnyOS.AnyCPU.Release\System.Dynamic.Runtime.Tests\default.netcoreapp1.1
set COMPlus_AltJit=*
set COMPlus_AltJitName=protojit.dll
RunTests.cmd c:\git\corefx\packages
```
**Output**
```
Using c:\git\coref2\packages as folder for resolving package dependencies.
Executing in C:\git\coref2\bin\tests\AnyOS.AnyCPU.Release\System.Dynamic.Runtime.Tests\default.netcoreapp1.1\
Hard linking dependent files...
Finished linking needed files, moving to running tests.
Running tests... Start time: 15:20:12.61
Command(s):
call CoreRun.exe xunit.console.netcore.exe System.Dynamic.Runtime.Tests.dll  -xml testResults.xml -notrait Benchmark=true -notrait category=nonnetcoreapp1.1tests  -notrait category=OuterLoop -notrait category=failing -notrait category=nonwindowstests
xUnit.net console test runner (32-bit .NET Core)
Copyright (C) 2014 Outercurve Foundation.

Discovering: System.Dynamic.Runtime.Tests
Discovered:  System.Dynamic.Runtime.Tests
Starting:    System.Dynamic.Runtime.Tests

Assert failure(PID 15164 [0x00003b3c], Thread: 8412 [0x20dc]): Assertion failed '!"We should never encounter a reference to a lclVar that has a zero refCnt."' in 'DynamicClass:CallSite.Target(ref,ref,ref,ref):ref' (IL size 162)

    File: c:\git\corecl2\src\jit\liveness.cpp Line: 55
    Image: C:\git\coref2\bin\tests\AnyOS.AnyCPU.Release\System.Dynamic.Runtime.Tests\default.netcoreapp1.1\CoreRun.exe

Finished running tests.  End time=15:25:55.95, Exit code = 123456789
```

### PR DESCRIPTION
**Pull request by pgavlin**
`fgMorphTree` may introduce additional lclVar references. Call
`lvaRecursiveIncRefCounts` in `fgMorphBlockStmt` to ensure that ref
counts are conservatively correct.

Fixes #8258.

### CODE REVIEW

**Comment by pgavlin** [comment]
@BruceForstall @CarolEidt @dotnet/jit-contrib PTAL.

**Comment by pgavlin** [comment]
Note that this change will overestimate lclVar ref counts for lclVars that appear in trees that are remorphed. If we wanted to be precise, we could decrement ref counts before morphing.

**Comment by CarolEidt** [comment]
Can you open an issue for this? I think we need a better strategy for lclVar ref counts, but at a minimum we should presumably distinguish the re-morphing case and decrement ref counts prior to re-morphing a statement.

---
In reply to: [262615753](https://github.com/dotnet/coreclr/pull/8284#issuecomment-262615753) [](ancestors = 262615753)

**Comment by pgavlin** [comment]
> Can you open an issue for this?

#8288

**Comment by CarolEidt** [comment]
Thanks @pgavlin

File: src/jit/morph.cpp [hunk 1]
```diff
@@ -15322,11 +15321,16 @@ bool Compiler::fgMorphBlockStmt(BasicBlock* block, GenTreePtr stmt DEBUGARG(cons
 
     if (!removedStmt)
     {
-        /* Have to re-do the evaluation order since for example
-         * some later code does not expect constants as op1 */
+        if (lvaLocalVarRefCounted)
+        {
+            // fgMorphTree may have introduced new lclVar references. Bump the ref counts if requested.
+            lvaRecursiveIncRefCounts(stmt->gtStmtExpr);
```
Comment (briansull) [inline src/jit/morph.cpp]: Can you also add some new logging here. to help us track down any potential fallout from this new change.
Like this:
#ifdef DEBUG
             printf("IExtra increment of LclVar refCounts for %s \n", msg);
#endif

**Pull request merged**